cmd/chart-up/main.go

```diff
@@ -0,0 +1,304 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/helm/pkg/chartutil"
+)
+
+var (
+	name    string
+	version string
+	gitSHA  string
+)
+
+const (
+	envCharts   = "CHARTS"
+	filePackage = "package.yaml"
+	fileChart   = "Chart.yaml"
+	fileValues  = "values.yaml"
+)
+
+const usage = `Version: %s
+%[2]s - rke2-charts version references update
+
+Usage: %[2]s [options] <chart name> [<version field>=<version value>]
+Options:
+  -h		show this help message
+  -c dir	rke2-charts directory (defaults: $CHARTS value or "rke2-charts")
+  -i		write changes into their respective files
+  -p		print resulting yaml files on STDOUT
+  -v		print version
+`
+
+func main() {
+	flag.Usage = func() {
+		w := os.Stderr
+		for _, arg := range os.Args {
+			if arg == "-h" {
+				w = os.Stdout
+				break
+			}
+		}
+		fmt.Fprintf(w, usage, version, name)
+	}
+
+	var (
+		vers, inPlace, printOut bool
+		chartsDir               string
+	)
+
+	chartDefault := "rke2-charts"
+	if c, ok := os.LookupEnv(envCharts); ok {
+		chartDefault = c
+	}
+
+	flag.BoolVar(&vers, "v", false, "print version")
+	flag.BoolVar(&inPlace, "i", false, "write changes into their respective files")
+	flag.BoolVar(&printOut, "p", false, "print resulting yaml files on STDOUT")
+	flag.StringVar(&chartsDir, "c", chartDefault, "rke2-charts directory")
+	flag.Parse()
+
+	if vers {
+		fmt.Fprintf(os.Stdout, "version: %s - git sha: %s\n", version, gitSHA)
+		return
+	}
+
+	args := flag.Args()
+
+	if len(args) == 0 {
+		exitErr("no chart provided")
+	}
+
+	chartName := flag.Arg(0)
+	chartPath := filepath.Join(chartsDir, "packages", chartName)
+
+	overrides := make(map[string]string, len(args))
+	for _, v := range args[1:] {
+		a := strings.Split(v, "=")
+
+		if len(a) < 2 {
+			continue
+		}
+
+		overrides[a[0]] = a[1]
+	}
+
+	err := filepath.Walk(chartPath, func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			if strings.Contains(path, "generated") {
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		var u yamlUpdater
+		switch filepath.Base(path) {
+		case fileChart:
+			u = newUpdater([]string{"appVersion", "version"})
+		case filePackage:
+			u = newUpdater([]string{"packageVersion"})
+		case fileValues:
+			u = newImageUpdater("tag")
+		default:
+			return nil
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		var b bytes.Buffer
+		if _, err := io.Copy(&b, f); err != nil {
+			return err
+		}
+
+		if err := u.Load(b.Bytes()); err != nil {
+			return err
+		}
+		u.Update(overrides)
+
+		if !(printOut || inPlace) {
+			fmt.Print(u)
+		}
+
+		if printOut {
+			fmt.Print(chartutil.ToYaml(u))
```

Comment:
you might want to look into using logrus to show the program logs for consistency with other tools written in go

Reply:
I think we haven't adopted logrus on ecm-distro-tools yet. What do you think of incorporating it @briandowns ?

Reply:
I'm ok with adding it as long as it adds quality or necessary information to the data being printed. If the data being printed can stand on it's own, then it's fine to leave it out. I'll defer to your and @galal-hussein judgement in this case.

Reply:
Ok, think it adds value on info/debug messages into STDERR. But for that line in specific I'd rather stick to fmt since it's the output we're looking for.